### PR TITLE
Add tests for reconcileInitialize

### PR DESCRIPTION
**Pull request by zoetrope**
Signed-off-by: zoetrope <a.ikezoe@gmail.com>

### CODE REVIEW

File: controllers/mysqlcluster_controller_test.go [hunk 1]
```diff
@@ -76,44 +112,537 @@ spec:
...
 			}
 
+			_, err = reconciler.createOrUpdateStatefulSet(ctx, reconciler.Log, cluster)
+			Expect(err).Should(HaveOccurred())
+		})
+
+	})
+
+	Context("CronJob", func() {
+		It("should create cron job", func() {
```
Comment (kfyharukz) [inline controllers/mysqlcluster_controller_test.go:556]: Some additional tests are needed, such as the CronJob has the correct command with the agent token.

File: controllers/mysqlcluster_controller_test.go [hunk 2]
```diff
@@ -76,44 +112,537 @@ spec:
...
-					return err
+				err = k8sClient.Get(ctx, client.ObjectKey{Name: moco.UniqueName(cluster), Namespace: cluster.Namespace}, cm)
+				Expect(err).ShouldNot(HaveOccurred())
+
+				Expect(cm.Data).Should(HaveKey(moco.MySQLConfName))
+				conf := cm.Data[moco.MySQLConfName]
+				Expect(conf).ShouldNot(ContainSubstring("innodb_buffer_pool_size"))
+			})
+
+			By("setting the size of 70% of the request", func() {
```
Comment (kfyharukz) [inline controllers/mysqlcluster_controller_test.go:248]: `70%` seems to be a magic number.  Should it be a constant?

File: controllers/mysqlcluster_controller_test.go [hunk 3]
```diff
@@ -76,44 +112,537 @@ spec:
...
+			}
+			Expect(claim).ShouldNot(BeNil())
+			Expect(claim.Spec.Resources.Requests.Storage().Value()).Should(BeNumerically("==", 1<<30))
+
+			isUpdated, err = reconciler.createOrUpdateStatefulSet(ctx, reconciler.Log, cluster)
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(isUpdated).Should(BeFalse())
+		})
+
+		It("should return error, when template does not contain mysqld container", func() {
```
Comment (kfyharukz) [inline controllers/mysqlcluster_controller_test.go:365]: This constraint doesn't appear in the docs.  We should write it in `crd_mysqlcluster.md`.

File: controllers/mysqlcluster_controller_test.go [hunk 4]
```diff
@@ -76,44 +112,537 @@ spec:
...
+						},
+					},
+				},
+			}
+
+			_, err := reconciler.createOrUpdateStatefulSet(ctx, reconciler.Log, cluster)
+			Expect(err).Should(HaveOccurred())
+		})
+
+		It("should return error, when template contains agent container", func() {
```
Comment (kfyharukz) [inline controllers/mysqlcluster_controller_test.go:383]: ditto

File: controllers/mysqlcluster_controller_test.go [hunk 5]
```diff
@@ -76,44 +112,537 @@ spec:
...
+			var dataClaim *corev1.PersistentVolumeClaim
+			for i, v := range sts.Spec.VolumeClaimTemplates {
+				if v.Name == "test-volume" {
+					testClaim = &sts.Spec.VolumeClaimTemplates[i]
+				}
+				if v.Name == mysqlDataVolumeName {
+					dataClaim = &sts.Spec.VolumeClaimTemplates[i]
+				}
+			}
+			Expect(testClaim).ShouldNot(BeNil())
```
Comment (kfyharukz) [inline controllers/mysqlcluster_controller_test.go:516]: The `test-volume` should be checked as having the correct value, shouldn't it?

**Review by kfyharukz — CHANGES_REQUESTED** [review]

File: controllers/mysqlcluster_controller_test.go [hunk 6]
```diff
@@ -76,44 +112,537 @@ spec:
...
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-primary", moco.UniqueName(cluster)), Namespace: namespace}, createdPrimaryService)
+			Expect(err).ShouldNot(HaveOccurred())
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-replica", moco.UniqueName(cluster)), Namespace: namespace}, createdReplicaService)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			Expect(createdPrimaryService.Spec.Ports).Should(HaveLen(2))
+			Expect(createdPrimaryService.Spec.Ports[0].Name).Should(Equal("mysql"))
+			Expect(createdPrimaryService.Spec.Ports[0].Port).Should(BeNumerically("==", moco.MySQLPort))
+
+			Expect(createdReplicaService.Spec.Ports).Should(HaveLen(2))
```
Comment (umezawatakeshi) [inline controllers/mysqlcluster_controller_test.go:594]: Is it intended to only check the length of "createdReplicaService" and not its content?

File: controllers/mysqlcluster_controller_test.go [hunk 7]
```diff
@@ -76,44 +112,537 @@ spec:
...
+
+			Expect(createdPrimaryService.Spec.Type).Should(Equal(corev1.ServiceTypeLoadBalancer))
+			Expect(createdReplicaService.Spec.Type).Should(Equal(corev1.ServiceTypeLoadBalancer))
+
 			Expect(createdPrimaryService.Spec.Ports).Should(HaveLen(2))
-			Expect(createdPrimaryService.Spec.Ports).Should(ConsistOf(expectedPorts))
+			Expect(createdPrimaryService.Spec.Ports[0].Name).Should(Equal("mysql"))
+			Expect(createdPrimaryService.Spec.Ports[0].Port).Should(BeNumerically("==", moco.MySQLPort))
 
 			Expect(createdReplicaService.Spec.Ports).Should(HaveLen(2))
```
Comment (umezawatakeshi) [inline controllers/mysqlcluster_controller_test.go:639]: ditto

**Review by kfyharukz — APPROVED** [review]
LGTM

**Pull request merged**